src/ioctls/common/kvm_vec.rs

```diff
@@ -0,0 +1,542 @@
+use std::mem;
+use std::mem::size_of;
+
+/// Errors associated with the KvmVec struct.
+#[derive(Debug, Clone)]
+pub enum Error {
+    /// The max size has been exceeded
+    SizeLimitExceeded,
+}
+
+/// Trait for accessing some properties of certain KVM structures that resemble an array.
+///
+/// The kvm API has many structs that resemble the following `MockKvmArray` structure:
+///
+/// # Example
+///
+/// ```
+/// extern crate kvm_bindings;
+/// use kvm_bindings::*;
+///
+/// use kvm_ioctls::{KvmArray, KvmVec};
+///
+/// const MAX_LEN: usize = 100;
+///
+/// #[repr(C)]
+/// #[derive(Default)]
+/// struct MockKvmArray {
+///     pub len: __u32,
+///     pub padding: __u32,
+///     pub entries: __IncompleteArrayField<__u32>,
+/// }
+///
+/// impl KvmArray for MockKvmArray {
+///     type Entry = u32;
+///
+///     fn len(&self) -> usize {
+///         self.len as usize
+///     }
+///
+///     fn set_len(&mut self, len: usize) {
+///         self.len = len as u32
+///     }
+///
+///     fn max_len() -> usize {
+///         MAX_LEN
+///     }
+///
+///     fn entries(&self) -> &__IncompleteArrayField<u32> {
+///         &self.entries
+///     }
+///
+///     fn entries_mut(&mut self) -> &mut __IncompleteArrayField<u32> {
+///         &mut self.entries
+///     }
+/// }
+/// ```
+#[allow(clippy::len_without_is_empty)]
+pub trait KvmArray {
+    /// The type of the __IncompleteArrayField entries
+    type Entry: PartialEq;
+
+    /// Get the array length
+    ///
+    fn len(&self) -> usize;
+
+    /// Get the array length as mut
+    ///
+    fn set_len(&mut self, len: usize);
+
+    /// Get max array length
+    ///
+    fn max_len() -> usize;
+
+    /// Get the array entries
+    ///
+    fn entries(&self) -> &__IncompleteArrayField<Self::Entry>;
+
+    /// Get the array entries as mut
+    ///
+    fn entries_mut(&mut self) -> &mut __IncompleteArrayField<Self::Entry>;
+}
+
+/// An adapter that helps in treating a KvmArray more like an actual `Vec`.
+///
+pub struct KvmVec<T: Default + KvmArray> {
+    // this variable holds the KvmArray structure. We use a `Vec<T>` To make the allocation
+    // large enough while still being aligned for `T`. Only the first element of `Vec<T>` will
+    // actually be used as a `T`. The remaining memory in the `Vec<T>` is for `entries`, which
+    // must be contiguous. Since the entries are of type `KvmArray::Entry` we must be careful to convert the
+    // desired capacity of the `KvmVec` from `KvmArray::Entry` to `T` when reserving or releasing memory.
+    kvm_array: Vec<T>,
+    // the number of elements of type `KvmArray::Entry` currently in the vec
+    len: usize,
+    // the capacity of the `KvmVec` measured in elements of type `KvmArray::Entry`
+    capacity: usize,
+}
+
+impl<T: Default + KvmArray> KvmVec<T> {
+    /// Returns the capacity required by kvm_array in order to hold the provided number of  `KvmArray::Entry`
+    ///
+    fn num_elements_to_kvm_array_len(num_elements: usize) -> usize {
```

Comment:
nit: I would rename this to kvm_vec_capacity. And also add a comment that it returns the kvm vec capacity required to hold `num_elements` of `KvmArray::Entry`.